File: commit/file.go
package commit

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"os"
	"sort"
	"syscall"

	"github.com/ava-labs/libevm/common"
	"github.com/ava-labs/libevm/core/types"
	"github.com/ava-labs/libevm/rlp"
)

// File is an opened, mmapped node file.
type File struct {
	m  []byte
	ft footer
}

// Open mmaps a node file written by Roll and checks its footer.
func Open(path string) (*File, error) {
	fd, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer fd.Close()
	st, err := fd.Stat()
	if err != nil {
		return nil, err
	}
	size := st.Size()
	if size < int64(len(magic))+footerSize {
		return nil, errCorrupt
	}
	m, err := syscall.Mmap(int(fd.Fd()), 0, int(size), syscall.PROT_READ, syscall.MAP_SHARED)
	if err != nil {
		return nil, err
	}
	ft, err := parseFooter(m[size-footerSize:])
	if err == nil && (string(m[:8]) != magic || ft.rootOff >= size || ft.idxOff < int64(len(magic)) ||
		ft.idxOff+int64(ft.idxN)*indexEntry != size-footerSize) {
		err = errCorrupt
	}
	if err != nil {
		syscall.Munmap(m)
		return nil, fmt.Errorf("%w: %s", err, path)
	}
	return &File{m: m, ft: ft}, nil
}

// Close unmaps the file.
func (f *File) Close() error { return syscall.Munmap(f.m) }

// Root is the state root the file was rolled at.
func (f *File) Root() common.Hash { return f.ft.root }

// UserData is the 32-byte field Roll was given.
func (f *File) UserData() [32]byte { return f.ft.user }

// NodeCount is the number of internal nodes in the file; KeyCount the number
// of leaves the roll hashed; Size the file size in bytes.
func (f *File) NodeCount() uint64 { return f.ft.nodes }
func (f *File) KeyCount() uint64  { return f.ft.keys }
func (f *File) Size() int64       { return int64(len(f.m)) }

// StorageRoot is the storage root of an account at the roll: EmptyRootHash
// when the account had no storage, and ok=false when the root is a leaf (one
// slot) rather than an internal node, in which case the hash is unknown to
// the file and the caller derives it from the flat row.
func (f *File) StorageRoot(acct common.Hash) (common.Hash, bool) {
	e := f.index(acct)
	if e == nil {
		return types.EmptyRootHash, false
	}
	return common.BytesToHash(e[32:64]), true
}

// index returns the index entry of acct, nil if absent.
func (f *File) index(acct common.Hash) []byte {
	idx := f.m[f.ft.idxOff : f.ft.idxOff+int64(f.ft.idxN)*indexEntry]
	n := int(f.ft.idxN)
	i := sort.Search(n, func(i int) bool { return bytes.Compare(idx[i*indexEntry:i*indexEntry+32], acct[:]) >= 0 })
	if i == n || !bytes.Equal(idx[i*indexEntry:i*indexEntry+32], acct[:]) {
		return nil
	}
	return idx[i*indexEntry : (i+1)*indexEntry]
}

// Node returns the RLP blob of the internal node at path (nibbles, as the
// trie package's Reader is asked) in the trie of owner (the zero hash for the
// account trie, else the account hash). ok=false when nothing is stored
// there: an absent, embedded, or leaf node. The blob aliases the mmap.
func (f *File) Node(owner common.Hash, path []byte) ([]byte, bool) {
	off := f.ft.rootOff
	if owner != (common.Hash{}) {
		e := f.index(owner)
		if e == nil {
			return nil, false
		}
		off = int64(binary.LittleEndian.Uint64(e[64:]))
	}
	for off != 0 {
		tag := f.m[off]
		n, k := binary.Uvarint(f.m[off+1:])
		blob := f.m[off+1+int64(k) : off+1+int64(k)+int64(n)]
		rest := f.m[off+1+int64(k)+int64(n):]
		if len(path) == 0 {
			return blob, true
		}
		switch tag {
		case tagBranch:
			for i := 0; i < int(path[0]); i++ {
				_, k := binary.Uvarint(rest)
				rest = rest[k:]
			}
			child, _ := binary.Uvarint(rest)
			off = int64(child)
			path = path[1:]
		case tagExt:
			content, _, _ := rlp.SplitList(blob)
			key, _, _ := rlp.SplitString(content)
			nib := compactToHex(key)
			if !bytes.HasPrefix(path, nib) {
				return nil, false
			}
			child, _ := binary.Uvarint(rest)
			off = int64(child)
			path = path[len(nib):]
		default:
			return nil, false
		}
	}
	return nil, false
}

File: rs/state/src/commit/file.rs
//! An opened, mmapped node file written by `roll`.
//!
//! File layout (all integers little-endian):
//!
//! header: magic (8)
//! nodes:  tag (1: branch, 2: extension), uvarint blob length, blob,
//!         then 16 uvarint child offsets (branch) or 1 (extension);
//!         0 means the child is absent, embedded in the parent blob, or a leaf
//! index:  indexCount x {account hash 32, storage root 32, root offset 8},
//!         sorted by account hash, one entry per account whose storage root
//!         is an internal node
//! footer: magic 8, version 4, root offset 8, root hash 32, node count 8,
//!         key count 8, index offset 8, index count 8, user data 32, crc32 4

use super::{compact_to_hex, err, uvarint, Result, FOOTER_SIZE, INDEX_ENTRY, MAGIC, TAG_BRANCH, TAG_EXT, VERSION};
use crate::keccak::EMPTY_ROOT;
use crate::rlp;
use crate::Hash;
use memmap2::Mmap;
use std::path::Path;

#[derive(Clone, Copy, Debug)]
pub struct Footer {
    pub root_off: u64,
    pub root: Hash,
    pub nodes: u64,
    pub keys: u64,
    pub idx_off: u64,
    pub idx_n: u64,
    pub user: [u8; 32],
}

pub fn parse_footer(b: &[u8]) -> Result<Footer> {
    if b.len() != FOOTER_SIZE || &b[..8] != MAGIC || u32::from_le_bytes(b[8..12].try_into().unwrap()) != VERSION {
        return err("commit: not a commit node file or corrupted footer");
    }
    if crc32fast::hash(&b[..FOOTER_SIZE - 4]) != u32::from_le_bytes(b[FOOTER_SIZE - 4..].try_into().unwrap()) {
        return err("commit: not a commit node file or corrupted footer");
    }
    let u64at = |o: usize| u64::from_le_bytes(b[o..o + 8].try_into().unwrap());
    Ok(Footer {
        root_off: u64at(12),
        root: b[20..52].try_into().unwrap(),
        nodes: u64at(52),
        keys: u64at(60),
        idx_off: u64at(68),
        idx_n: u64at(76),
        user: b[84..116].try_into().unwrap(),
    })
}

pub struct File {
    m: Mmap,
    ft: Footer,
}

impl File {
    /// Mmaps a node file written by roll and checks its footer.
    pub fn open(path: &Path) -> Result<File> {
        let fd = std::fs::File::open(path)?;
        let size = fd.metadata()?.len();
        if size < (MAGIC.len() + FOOTER_SIZE) as u64 {
            return err(format!("commit: not a commit node file or corrupted footer: {}", path.display()));
        }
        let m = unsafe { Mmap::map(&fd)? };
        let ft = parse_footer(&m[size as usize - FOOTER_SIZE..]).map_err(|e| super::Error(format!("{e}: {}", path.display())))?;
        if &m[..8] != MAGIC || ft.root_off >= size || ft.idx_off < MAGIC.len() as u64 || ft.idx_off + ft.idx_n * INDEX_ENTRY as u64 != size - FOOTER_SIZE as u64 {
            return err(format!("commit: not a commit node file or corrupted footer: {}", path.display()));
        }
        Ok(File { m, ft })
    }

    /// The state root the file was rolled at.
    pub fn root(&self) -> Hash {
        self.ft.root
    }
    pub fn user_data(&self) -> [u8; 32] {
        self.ft.user
    }
    pub fn node_count(&self) -> u64 {
        self.ft.nodes
    }
    pub fn key_count(&self) -> u64 {
        self.ft.keys
    }
    pub fn size(&self) -> u64 {
        self.m.len() as u64
    }
    pub fn footer(&self) -> &Footer {
        &self.ft
    }

    /// The storage root of an account at the roll: EMPTY_ROOT when the
    /// account had no storage, and false when the root is a leaf (one slot)
    /// rather than an internal node, in which case the hash is unknown to
    /// the file and the caller derives it from the flat row.
    pub fn storage_root(&self, acct: &Hash) -> (Hash, bool) {
        match self.index(acct) {
            None => (EMPTY_ROOT, false),
            Some(e) => (e[32..64].try_into().unwrap(), true),
        }
    }

    fn index(&self, acct: &Hash) -> Option<&[u8]> {
        let idx = &self.m[self.ft.idx_off as usize..self.ft.idx_off as usize + self.ft.idx_n as usize * INDEX_ENTRY];
        let n = self.ft.idx_n as usize;
        let (mut lo, mut hi) = (0usize, n);
        while lo < hi {
            let m = (lo + hi) / 2;
            if &idx[m * INDEX_ENTRY..m * INDEX_ENTRY + 32] < &acct[..] {
                lo = m + 1;
            } else {
                hi = m;
            }
        }
        if lo == n || &idx[lo * INDEX_ENTRY..lo * INDEX_ENTRY + 32] != &acct[..] {
            return None;
        }
        Some(&idx[lo * INDEX_ENTRY..(lo + 1) * INDEX_ENTRY])
    }

    /// The RLP blob of the internal node at path (nibbles) in the trie of
    /// owner (the zero hash for the account trie, else the account hash).
    /// None when nothing is stored there: an absent, embedded, or leaf node.
    pub fn node(&self, owner: &Hash, mut path: &[u8]) -> Option<&[u8]> {
        let mut off = self.ft.root_off;
        if owner != &[0u8; 32] {
            let e = self.index(owner)?;
            off = u64::from_le_bytes(e[64..72].try_into().unwrap());
        }
        while off != 0 {
            let o = off as usize;
            let tag = self.m[o];
            let (n, k) = uvarint(&self.m[o + 1..]);
            let blob = &self.m[o + 1 + k..o + 1 + k + n as usize];
            let mut rest = &self.m[o + 1 + k + n as usize..];
            if path.is_empty() {
                return Some(blob);
            }
            match tag {
                TAG_BRANCH => {
                    for _ in 0..path[0] {
                        let (_, k) = uvarint(rest);
                        rest = &rest[k..];
                    }
                    off = uvarint(rest).0;
                    path = &path[1..];
                }
                TAG_EXT => {
                    let (content, _) = rlp::split_list(blob)?;
                    let (key, _) = rlp::split_string(content)?;
                    let nib = compact_to_hex(key);
                    if !path.starts_with(&nib) {
                        return None;
                    }
                    off = uvarint(rest).0;
                    path = &path[nib.len()..];
                }
                _ => return None,
            }
        }
        None
    }
}
